Improved error messages for `max_unpool{}d` operators

krshrimali opened the pull request:
~As per the title, this PR adds OpInfos for `max_unpoolNd` operators. There are a few TODOs:~

* [x] Improve error messages for the rest of the functions in the CUDA file for the un-pooling operators.
~* [x] Raise issues for the failures, and provide descriptions.~

~Note to the reviewers: I'll add descriptions and reasons for the skips, I'm not totally sure about them, hence the skips for now.~

cc: @mruberry @saketh-are

krshrimali committed:
Unpool1d OpInfo, minor fix in the error msg

krshrimali committed:
For max_unpool2d, also refactor error messages

krshrimali committed:
Add skip for test_variant_consistency_jit

krshrimali committed:
Lint fixes

krshrimali committed:
More fixes in error messages, for max unpooling2d and unpooling3d

krshrimali committed:
Add OpInfo for max_unpool3d

krshrimali committed:
bfloat16 not supported on CUDA for max_unpool3d

krshrimali committed:
Error messages fixed in CUDA kernel for unpooling

krshrimali committed:
Merge branch 'origin/max_unpoolNd' of github.com:krshrimali/pytorch into origin/max_unpoolNd

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into origin/max_unpoolNd

krshrimali committed:
Reference issues, bfloat16 not supported on CUDA for max_unpoolNd?

krshrimali committed:
Merge branch 'origin/max_unpoolNd' of github.com:krshrimali/pytorch into origin/max_unpoolNd

krshrimali commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8685,6 +8796,49 @@ def ref_pairwise_distance(input1, input2):
+               # TODO: Reference issue here
+               DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_fn_grad', dtypes=(torch.float64,)),
+               DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_fn_gradgrad', dtypes=(torch.float64,)),
+               DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit', dtypes=(torch.float32,)),
```
Have raised an issue here: https://github.com/pytorch/pytorch/issues/67659

krshrimali committed:
Issue for JIT error

krshrimali committed:
Merge branch 'master' into origin/max_unpoolNd

krshrimali committed:
remove grad_fn history from the tensor, minor typo fixes in error messages

krshrimali commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2990,6 +2990,117 @@ def generator():
+            output_pool, indices = max_pool1d(input_tensor)
+            if output_size:
+                kwargs.update({'output_size': input_tensor.size()})
+            yield SampleInput(output_pool.detach().requires_grad_(requires_grad), args=(indices,), kwargs=kwargs)
```
This `output_pool.detach().requires_grad_(requires_grad)` is to make sure that there is no `grad_fn` history recorded for `output_pool` tensor. If not done, this will fail the `test_noncontiguous_samples` test:

https://github.com/pytorch/pytorch/blob/06d1be2447ad2041336ee93c435345a4f371b9a9/test/test_ops.py#L228

Good to have that assertion in the test, else I would have unknowingly forgotten to do this.

cc: @mruberry

krshrimali reviewed (commented):
Hi @jbschlosser - I was wondering if you would mind reviewing the changes in the PR for error messages (if you have the time)? I personally feel it is good to print sizes in the error messages for which it failed, and what was expected. I'm looking forward to your comments and any feedback.

krshrimali commented:
Gentle ping, @mruberry / @saketh-are - can you guys please help review this PR whenever you get the time? Thanks! :)

mruberry commented:
Thanks for the ping, @krshrimali! @saketh-are - would you review these additions?

saketh-are commented on aten/src/ATen/native/MaxUnpooling.cpp:
```diff
@@ -14,16 +14,16 @@ Tensor& max_unpooling2d_forward_out_cpu(
   TORCH_CHECK(
       self_.sizes() == indices_.sizes(),
-      "Shape of indices should match shape of input");
+      "Shape of indices (", indices_.sizes(), ") should match shape of input (", self_.sizes(), ")");
```
Moving `indices_.sizes()` to the end of this message with a "but got" would improve consistency with other error messages and may improve readability. Presently, this error mesage reverses the order of the "expected" value and the "actual" value when compared to other error messages.

krshrimali replied:
This makes sense, consistency in error messages is always great to have, I agree! Thanks for the suggestion, @saketh-are - I've made the requested changes, hope they look good.

saketh-are commented on aten/src/ATen/native/MaxUnpooling.cpp:
```diff
@@ -77,20 +77,19 @@ static void max_unpooling3d_shape_check(
   TORCH_CHECK(
       input.sizes() == indices.sizes(),
-      "Shape of indices should match shape of input");
+      "Shape of indices (", indices.sizes(), ") should match shape of input (", input.sizes(), ")");
```
Same here

krshrimali replied:
Done. :)

saketh-are commented on aten/src/ATen/native/MaxUnpooling.cpp:
```diff
@@ -197,10 +196,11 @@ Tensor& max_unpooling2d_backward_out_cpu(const Tensor& grad_output_,
   TORCH_CHECK(
-      self.sizes() == indices_.sizes(), "Input shape must match indices shape");
-  TORCH_CHECK(output_size.size() == 2, "Output size must be 2");
+      self.sizes() == indices_.sizes(), "Input shape (", self.sizes(), ") must match indices shape (",
```
Same here

krshrimali replied:
Done. :)

saketh-are commented on aten/src/ATen/native/cpu/MaxUnpoolKernel.cpp:
```diff
@@ -149,7 +149,7 @@ void cpu_max_unpool_channels_last(
     AT_ERROR("Found an invalid max index: ", optional_error_index.value(),
         " (output volumes are of size ", output_height,
-        "x", output_width);
+        "x", output_width, ")");
```
Nice catch

krshrimali replied:
Thanks! :)

saketh-are commented on aten/src/ATen/native/cuda/MaxUnpooling.cu:
```diff
@@ -122,14 +122,13 @@ Tensor& max_unpooling2d_forward_out_cuda(const Tensor& self_,
   TORCH_CHECK(
       self_.sizes() == indices_.sizes(),
-      "Shape of input must match shape of indices");
+      "Shape of indices (", indices_.sizes(), ") should match shape of input (", self_.sizes(), ")");
```
Another instance

krshrimali replied:
Done. :)

saketh-are commented on aten/src/ATen/native/cuda/MaxUnpooling.cu:
```diff
@@ -204,23 +203,22 @@ static void max_unpooling3d_shape_check(
   TORCH_CHECK(
       input.sizes() == indices.sizes(),
-      "Shape of indices should match shape of input");
+      "Shape of indices (", indices.sizes(), ") should match shape of input (", input.sizes(), ")");
```
One more

krshrimali replied:
Done. :)

saketh-are commented on aten/src/ATen/native/cuda/MaxUnpooling.cu:
```diff
@@ -386,7 +384,7 @@ at::Tensor& max_unpooling2d_backward_out_cuda(const Tensor& grad_output_,
   TORCH_CHECK(
       indices_.scalar_type() == at::ScalarType::Long,
-      "elements in indices should be type int64");
+      "elements in indices should be type int64 but got tyoe: ", indices_.scalar_type());
```
nit: typo

krshrimali replied:
Done. :) Sorry for the typo.

saketh-are commented on aten/src/ATen/native/cuda/MaxUnpooling.cu:
```diff
@@ -401,9 +399,9 @@ at::Tensor& max_unpooling2d_backward_out_cuda(const Tensor& grad_output_,
   TORCH_CHECK(
       self_.sizes() == indices_.sizes(),
-      "Input should have same shape as indices");
+      "Input shape (", self_.sizes(), ") must match indices shape (", indices_.sizes(), ")");
```
Another instance

krshrimali replied:
Fixed. :)

saketh-are commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2990,6 +2990,117 @@ def generator():
+                "padding": max_unpool1d.padding,
+            }
+
+            output_pool, indices = max_pool1d(input_tensor)
```
Using `MaxPool1d` to produce the inputs for `MaxUnpool1d` is a wonderful choice

saketh-are commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2990,6 +2990,117 @@ def generator():
+                                            dilation=dilation, return_indices=True)
+
+            kwargs = {
+                "kernel_size": max_unpool1d.kernel_size,
```
Is there a reason we need to declare `max_unpool1d` and get these kwargs back out from it instead of using them directly?

krshrimali replied:
Thanks for the question, @saketh-are. Using `nn.MaxUnpool1d` allows to test default behavior of these kwargs, and what happens when passed `None`. But for this PR, I've removed OpInfo changes since these were incorporated in a different PR by @nikitaved. (I'll link to the PR soon)

krshrimali replied:
Update: looks like I was wrong. The OpInfos for `max_unpool{}d` aren't there yet. @nikitaved - Do you plan to work on them? Or can I add the OpInfos in this PR? Please let me know if I missed any open PR on this. (same question for `max_pool{}d_with_indices`)

nikitaved replied:
I think it makes sense to keep OpInfo tests from https://github.com/pytorch/pytorch/pull/68625 as they do not skip gradchecks. Waiting on @mruberry to approve them.

krshrimali replied:
Ah, got it. Sorry, I missed it. Thanks, @nikitaved. Quick Question: are you working on OpInfos for `max_pool{}d_with_indices` as well?

nikitaved replied:
Not yet, no.

saketh-are reviewed (commented):
@krshrimali these OpInfos are looking good! Thanks also for the great updates to the error messages. I left some comments inline for your consideration. Look forward to seeing what you think.

krshrimali committed:
New format of error messages as per review

krshrimali committed:
Not have OpInfos

krshrimali commented:
Thanks, @saketh-are for reviewing this PR. The OpInfos were added in another PR by @nikitaved, so I've removed those changes from this PR, and have limited this to improvement in error messages.

I've addressed your review comments, please let me know if they look good to you and if there is anything else that can be improved. :) Apologies for the delay in response, and thank you for your patience. :)

mruberry commented:
@albanD, @saketh-are is still out -- would you do a quick final review and shepherd this PR in?

albanD commented:
Sure

albanD commented on aten/src/ATen/native/MaxUnpooling.cpp:
```diff
@@ -14,16 +14,16 @@ Tensor& max_unpooling2d_forward_out_cpu(
   TORCH_CHECK(
       self_.sizes() == indices_.sizes(),
-      "Shape of indices should match shape of input");
+      "Expected shape of indices to be: ", self_.sizes(), " but got: ", indices_.sizes());
```
nit: we lost the information that the expected size is the input's size no?

krshrimali replied:
Ah, right, good catch Alban. Thanks.

The error message now looks like below:

```python
RuntimeError: Expected shape of indices to be same as that of the input tensor ([1, 1, 2, 2]) but got indices tensor with shape: [2, 2, 1, 1]
```

Do you think this is better?

albanD replied:
Yep thanks!

albanD commented on aten/src/ATen/native/MaxUnpooling.cpp:
```diff
@@ -77,20 +77,19 @@ static void max_unpooling3d_shape_check(
       (input.ndimension() == 4 || input.ndimension() == 5),
-      "Input to max_unpooling3d should be a 4d or 5d Tensor",
-      input.sizes());
+      "Input to max_unpooling3d should be a 4d or 5d Tensor, but got a tensor with dim ", input.ndimension());
```
nit dim -> X dimensions

krshrimali replied:
My bad, thanks Alban for pointing this out. Should be fixed now.

albanD reviewed (commented):
I'm surprised there isn't a single test changed when so many error message are updated?

nikitaved commented:
@albanD , there is a stale https://github.com/pytorch/pytorch/pull/68625 that tests this function more comprehensively.

albanD commented:
Ok Sounds good. @nikitaved anything special blocking this OpInfo PR? Shall we revive it?

@krshrimali could you rebase on latest master and do the minor updates listed above please? This will be ready to land after that!

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into origin/max_unpoolNd

krshrimali committed:
Address review, revise error messages

krshrimali commented:
Thanks, @albanD for reviewing the PR. It should be ready for another look.

nikitaved commented:
@albanD , thanks! Nothing really, it is already good as it is.

albanD reviewed (approved):
Perfect!
Thanks for the update.
